presto-main/src/main/java/com/facebook/presto/operator/scalar/UnicodeUtil.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.scalar;
+
+import io.airlift.slice.Slice;
+
+/**
+ * Ideas for fast counting based on
+ * <a href="http://www.daemonology.net/blog/2008-06-05-faster-utf8-strlen.html">Even faster UTF-8 character counting</a>.
+ */
+final class UnicodeUtil
+{
+    private UnicodeUtil()
+    {
+    }
+
+    private static final int TOP_MASK32 = 0x80808080;
+    private static final long TOP_MASK64 = 0x8080808080808080L;
+
+    private static final int ONE_MULTIPLIER32 = 0x01010101;
+    private static final long ONE_MULTIPLIER64 = 0x0101010101010101L;
+
+    /**
+     * Counts the code points within UTF8 encoded slice.
+     */
+    static int countCodePoints(final Slice string)
+    {
+        return countCodePoints(string, string.length());
+    }
+
+    /**
+     * Counts the code points within UTF8 encoded slice up to {@code end}.
+     */
+    static int countCodePoints(final Slice string, int end)
+    {
+        //
+        // Quick exit if empty string
+        if (end == 0) {
+            return 0;
+        }
+
+        int count = 0;
+        int i = 0;
+        //
+        // Length rounded to 8 bytes
+        final int length8 = end & 0x7FFFFFF8;
+        for (; i < length8; i += 8) {
+            //
+            // Fetch 8 bytes as long
+            long i64 = string.getLong(i);
+            //
+            // Count bytes which are NOT the start of a code point
+            i64 = ((i64 & TOP_MASK64) >> 7) & (~i64 >> 6);
```

Comment:
In general, you're right, but in this case it does not matter.

If the upper 7 bits of i64 look like
`1xxx xxxx`
 then `i64 & TOP_MASK64) >> 7` evaluates to
`1111 1111`
and `~i64 >> 6`evaluates to
`0000 000x`

After the `and`, the upper bits are ignored. To avoid confusion in the future I can change to shift without sign extension.